File: plot/src/main/java/net/fwitz/math/plot/binary/complex/color/DomainColoringContour.java
package net.fwitz.math.plot.binary.complex.color;

import net.fwitz.math.binary.complex.Complex;

import java.awt.*;

import static java.awt.Color.getHSBColor;
import static java.lang.Math.floor;

public class DomainColoringContour implements ComplexColorFunction {
    private static final double TWO_PI = Math.PI * 2.0;
    private static final double LN_2_RECIP = 1.0 / Math.log(2.0);

    private static float hue(Complex z) {
        double hue = z.arg() / TWO_PI;
        return (float) (hue - floor(hue));
    }

    private static double logScaleMantissa(Complex z) {
        double m = z.logabs() * LN_2_RECIP;
        return m - floor(m);
    }

    private static float brightness(double k) {
        //return (float) (0.4 + 0.6 * k);
        return (float) Math.pow(k, 0.2);
    }


    public Color apply(Complex c, Complex z) {
        double k = logScaleMantissa(z);
        float hue = hue(z);
        float val = brightness(k);
        return getHSBColor(hue, 1.0f, val);
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/binary/complex/color/DomainColoringContour.kt
package net.fwitz.math.plot.binary.complex.color

import net.fwitz.math.binary.complex.Complex
import java.awt.Color
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.pow

object DomainColoringContour : ComplexColorFunction {
    private const val TWO_PI = Math.PI * 2.0
    private val LN_2_RECIP = 1.0 / ln(2.0)

    private fun hue(z: Complex): Float {
        val hue: Double = z.arg / TWO_PI
        return (hue - floor(hue)).toFloat()
    }

    private fun logScaleMantissa(z: Complex): Double {
        val m: Double = z.logabs * LN_2_RECIP
        return m - floor(m)
    }

    private fun brightness(k: Double): Float {
        //return (float) (0.4 + 0.6 * k);
        return k.pow(0.2).toFloat()
    }

    override fun invoke(c: Complex, z: Complex): Color {
        val k = logScaleMantissa(z)
        val hue = hue(z)
        val brt = brightness(k)
        return Color.getHSBColor(hue, 1.0f, brt)
    }
}